File: pyapp/src/pyapp/infrastructure/data_repository.py
from __future__ import annotations
from typing import Dict, Generic, TypeVar, Optional, List
import time
import secrets

T = TypeVar("T")

class DataRepository(Generic[T]):
    def __init__(self) -> None:
        self._store: Dict[str, Dict[str, T]] = {}

    def _bucket(self, type_: str) -> Dict[str, T]:
        k = type_.lower()
        if k not in self._store:
            self._store[k] = {}
        return self._store[k]

    async def create(self, type_: str, item: T) -> T:
        id_ = f"{int(time.time()*1000)}-{secrets.token_hex(6)}"
        full = dict(item)  # type: ignore
        full["id"] = id_
        self._bucket(type_)[id_] = full  # type: ignore
        return full  # type: ignore

    async def get(self, type_: str, id_: str) -> Optional[T]:
        return self._bucket(type_).get(id_)  # type: ignore

    async def delete(self, type_: str, id_: str) -> bool:
        return self._bucket(type_).pop(id_, None) is not None

    async def list(self, type_: str, q: str = "", page: int = 1, size: int = 20) -> List[T]:
        b = self._bucket(type_)
        arr = list(b.values())
        q = (q or "").lower()
        filtered = [it for it in arr if q in str(it).lower()] if q else arr
        page = max(1, page)
        size = max(1, min(100, size))
        start = (page - 1) * size
        return filtered[start : start + size]  # noqa: E203

data_repo: DataRepository[dict] = DataRepository()

File: apps/api/src/shared/data/data.repository.ts
export type Item = { id: string; [k: string]: any }

export class DataRepository<T extends { id: string }> {
    private store = new Map<string, Map<string, T>>()

    private bucket(type: string) {
        const k = type.toLowerCase()
        if (!this.store.has(k)) this.store.set(k, new Map())
        return this.store.get(k)!
    }

    async create(type: string, item: Omit<T, 'id'>): Promise<T> {
        const id = `${Date.now()}-${Math.random().toString(36).slice(2)}`
        const full = { ...(item as any), id } as T
        this.bucket(type).set(id, full)
        return full
    }

    async get(type: string, id: string): Promise<T | null> {
        return this.bucket(type).get(id) || null
    }

    async update(type: string, id: string, patch: Partial<T>): Promise<T | null> {
        const b = this.bucket(type)
        const cur = b.get(id)
        if (!cur) return null
        const next = { ...cur, ...patch, id } as T
        b.set(id, next)
        return next
    }

    async delete(type: string, id: string): Promise<boolean> {
        return this.bucket(type).delete(id)
    }

    async list(type: string, opts?: { q?: string; page?: number; size?: number }): Promise<T[]> {
        const b = this.bucket(type)
        const arr = Array.from(b.values())
        const q = (opts?.q || '').toLowerCase()
        const filtered = q ? arr.filter((it) => JSON.stringify(it).toLowerCase().includes(q)) : arr
        const page = Math.max(1, opts?.page || 1)
        const size = Math.max(1, Math.min(100, opts?.size || 20))
        const start = (page - 1) * size
        return filtered.slice(start, start + size)
    }
}

export const dataRepo = new DataRepository<Item>()
